Add support for no-op concat with padded output

nandesuka opened the pull request:
Add support for no-op concat with padded output

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

nandesuka commented:
@pytorchbot label "topic: not user facing"

blaine-rister commented on test/inductor/test_torchinductor_codegen_config_overrides.py:
```diff
@@ -108,6 +108,50 @@ def func(a, b):
+        ((32, 30), 64), 
+        ]
+    )
+    def test_concat_alignments(self, shape, alignment_bytes):
```
Nice test!

blaine-rister replied:
Can I suggest moving this test to this module? https://github.com/pytorch/pytorch/blob/main/test/inductor/test_padding.py

That contains a lot of similar padding tests.

blaine-rister commented on test/inductor/test_torchinductor_codegen_config_overrides.py:
```diff
@@ -108,6 +108,50 @@ def func(a, b):
+       
+        output_shape = (shape[0] + shape[0], shape[1])
+        output_stride = x.stride()
+        output_line = f"buf3 = empty_strided_cpu({output_shape}, {output_stride}, torch.float32)"
```
If the output code is this specific, you may want to consider using `expecttest` asserts. Those are a bit easier to maintain because you can automatically update the code with `EXPECTTEST_ACCEPT=1`. Here's an example. https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_strided_blocks.py#L350

More info on `expecttest`:
https://github.com/pytorch/expecttest

blaine-rister commented on torch/_inductor/ir.py:
```diff
@@ -4634,6 +4634,8 @@ def create(cls, inputs, dim):  # type: ignore[no-untyped-def]
 
         output_stride = FlexibleLayout.contiguous_strides(new_size)
+        # Ensure the output stride matches the alignment requirements
+        output_stride = Layout._pad_strides(output_stride, new_size, inputs[0].dtype)
```
I'm wondering if padded concats have any perf drawbacks. It seems like padding could be beneficial on GPUs to avoid bank conflicts, but I'm not sure if there's a reason the existing code specifically asks for contiguous strides. It might be good to double check for perf regressions by running some benchmarks.

You can kick off an A100 perf run here. https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml

It might also be useful to write a microbenchmark showing whether this impacts perf on programs that only contain concats. If you run Inductor in debug mode, it will automatically generate a benchmarking script. https://pytorch.org/tutorials/intermediate/inductor_debug_cpu.html#get-more-logging-information

cc @shunting314  @eellison

eellison replied:
If we have a no-op concat like:
```
def foo(a, b, x, y):
     return torch.cat((a @ b, x @ y))
```

Then padding the stride here I think will be perf-negative, bc we will no longer be able to plan this. The inputs were dense but our stride is not.

I think we want to do some checking to avoid padding in this case.

nandesuka replied:
Perf run completed. https://github.com/pytorch/pytorch/actions/runs/13567211431

blaine-rister reviewed (commented):
This mostly LGTM. I left a few small nits.

Before accepting, it would be good to see some GPU benchmarks to confirm this is at least perf neutral. (It might actually be perf positive if it avoids bank conflicts.)

cc @shunting314  in case you want to suggest some representative test programs.

blaine-rister commented on torch/_inductor/ir.py:
```diff
@@ -4634,6 +4634,8 @@ def create(cls, inputs, dim):  # type: ignore[no-untyped-def]
             offsets_end.append(new_size[dim])
 
         output_stride = FlexibleLayout.contiguous_strides(new_size)
```
On second thought, it seems like we should choose whether to pad based on `config.comprehensive_padding`. That would be more consistent with how `FlexibleLayout` usually chooses padding.

eellison replied:
Also, as @blaine-rister says youre skipping the `should_pad_strides` path

blaine-rister commented on test/inductor/test_torchinductor_codegen_config_overrides.py:
```diff
@@ -108,6 +108,50 @@ def func(a, b):
+                "max_pointwise_cat_inputs" : 1,
+        })
+       
+        output_shape = (shape[0] + shape[0], shape[1])
```
Can we also test what happens when `config.pad_outputs == False`? AFAIK Inductor will try to match eager mode strides (i.e., not pad) when that happens.

eellison reviewed (commented):
cc @shunting314 for thoughts too

shunting314 commented:
Looks good to me. +1 to guard the code behind the config.comprehensive_padding .

nandesuka committed:
Add support for no-op concat with padded output (#146866)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/146866

per title

Test Plan: buck2 run fbcode//mode/opt fbcode//mode/inplace fbcode//mode/mtia //mtia/compiler/graph_compiler/test_library:afg_inductor_test -- -k _ig_concat_

Reviewed By: shunting314, yelite

Differential Revision: D68918810
